### PR TITLE
W-11410770 Ensue that ObjectStore is always created to fix - oauth dance authorize getting stuck when using custom Redis object   store 

### PR DESCRIPTION
**Pull request by mule-pachanady**
Oauth dance authorize is stuck when using custom object store which uses Redis Cache

CCE noticed the following upon debug

Debuging the app I notice that when trying to obtain the object store in order to store the token I am getting a "ObjectStore 'object_store' is not defined".
protected Function<OAuthConfig, ObjectStore> buildObjectStoreLocator() {

  return config -> {

   Optional<OAuthObjectStoreConfig> storeConfig = config.getStoreConfig();

   String storeName = storeConfig

     .map(OAuthObjectStoreConfig::getObjectStoreName)

     .orElse(BASE_PERSISTENT_OBJECT_STORE_KEY);

   **return objectStoreManager.getObjectStore(storeName);**   Changed to-----> getOrCreateObjectStore

  };

 }

I deployed the provided manually and also created the Slack WS with Carlos, after the fix the app is starting and token is retrieved successfully upon accessing https //localhost/8081/activate.

### CODE REVIEW

**Comment by mule-pachanady** [comment]
@fsgonz  Please take a look tomorrow

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandler.java [hunk 1]
```diff
@@ -122,8 +129,14 @@ protected Function<OAuthConfig, ObjectStore> buildObjectStoreLocator() {
       String storeName = storeConfig
           .map(OAuthObjectStoreConfig::getObjectStoreName)
           .orElse(BASE_PERSISTENT_OBJECT_STORE_KEY);
-
-      return objectStoreManager.getObjectStore(storeName);
+      ObjectStore os = objectStoreManager.getOrCreateObjectStore(storeName, ObjectStoreSettings
+          .builder()
+          .entryTtl(0L)
```
Comment (fsgonz) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandler.java:134]: What is the reason for using this default settings?

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  }
+
+
+  @After
+  public void after() throws MuleException {
+    schedulerService.stop();
+  }
+
+  @Override
+  protected Map<String, Object> getStartUpRegistryObjects() {
```
Comment (fsgonz) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java:143]: Why is this needed?

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    assertThat(objectStore.allKeys(), is(OBJECT_STORE_ENTRIES));
+    verify(storeManager, times(1)).getOrCreateObjectStore(anyString(), any(ObjectStoreSettings.class));
+  }
+
+  @Rule
+  public TemporaryFolder tempWorkDir = new TemporaryFolder();
+
+  @Before
+  public void setup() throws ObjectStoreException {
+    schedulerService = new SimpleUnitTestSupportSchedulerService();
```
Comment (fsgonz) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java:112]: why is this needed?

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    muleContext = mock(MuleContextWithRegistry.class);
```
Comment (fsgonz) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java:113]: why is the muleContext needed?

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    when(configurationProperties.resolveStringProperty(OCS_PLATFORM_AUTH_URL)).thenReturn(of(PLATFORM_AUTH_URL));
+    when(configurationProperties.resolveStringProperty(OCS_API_VERSION)).thenReturn(empty());
+    when(configurationProperties.resolveStringProperty(OCS_API_VERSION)).thenReturn(empty());
+  }
+
+
+  @Override
+  protected void doTearDown() throws Exception {
+    oauthHandler.stop();
+    currentMuleContext.set(null);
```
Comment (fsgonz) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java:133]: What is this needed for?

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.runtime.connectivity.oauth;
```
Comment (fsgonz) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java:7]: Try to make this test minimal.

**Review by fsgonz — CHANGES_REQUESTED** [review]
Can we add some kind of integration test with a minimal app (with a config file) to make it clear the scenario under which this failed?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandler.java (hunk 1 above)
Comment (mule-pachanady) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandler.java:134]: I just wanted to go with some defaults and don't want to add more system properties, may be will change to ObjectStoreSettings.builder().persistent(true).build()

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java (hunk 2 above)
Comment (mule-pachanady) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java:143]: Made many changes, but not needed now

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java (hunk 4 above)
Comment (mule-pachanady) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java:113]: No longer needed

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java (hunk 5 above)
Comment (mule-pachanady) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java:133]: Shouldn't matter much for tests, but we have consistently done this in many tests

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java (hunk 6 above)
Comment (mule-pachanady) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java:7]: Will Try Thanks

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java (hunk 5 above)
Comment (fsgonz) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java]: Leave the test as minimal as needed. It is usually needed in other tests.

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java (hunk 2 above)
Comment (fsgonz) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java]: If not needed, remove it from the tests.

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java (hunk 3 above)
Comment (fsgonz) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandlerTestCase.java]: Same here. I think the scheduler service is not used in the test.

**Review by fsgonz — CHANGES_REQUESTED** [review]
Add an integration test with an app for the use case.

**Comment by mule-pachanady** [comment]
Im looking at adding a Intg test

**Comment by mule-pachanady** [comment]
--validate

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandler.java [hunk 7]
```diff
@@ -47,6 +52,8 @@
  */
 public abstract class OAuthHandler<Dancer> implements Lifecycle {
 
+  public static final ThreadLocal<MuleContext> currentMuleContext = new ThreadLocal<>();
```
Comment (fsgonz) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandler.java:55]: I think this is not used. What is the purpose of this?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandler.java [hunk 8]
```diff
@@ -47,6 +52,8 @@
...
+  private static final boolean DEFAULT_PERSISTENCE_SETTING = true;
```
Comment (fsgonz) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthHandler.java:56]: This is not used as default.

**Review by fsgonz — CHANGES_REQUESTED** [review]

**Review by fsgonz — APPROVED** [review]

**Review by fsgonz — APPROVED** [review]
